File: sdk/python/tensorfeed/langchain.py
"""TensorFeed LangChain integration.

Optional extra. Install with::

    pip install 'tensorfeed[langchain]'

Provides:
    * ``TensorFeedNewsTool``    - LangChain tool wrapping ``/api/news``
    * ``TensorFeedStatusTool``  - tool wrapping ``/api/status``
    * ``TensorFeedRoutingTool`` - tool wrapping the free ``/api/preview/routing``
    * ``TensorFeedAttentionTool`` - tool wrapping ``/api/attention``
    * ``TensorFeedHarnessesTool`` - tool wrapping ``/api/harnesses``
    * ``TensorFeedNewsLoader``  - LangChain document loader for news articles
    * ``tensorfeed_tools()``    - convenience: free tools, optionally premium
    * ``tensorfeed_premium_tools()`` - the paid catalog (1 credit each)

The premium tools never move funds. Called without a credit token they
return a clear guidance string instead of paying or raising. Provisioning
credits stays an explicit operator action; see ``_premium_tools``.

Usage::

    from langchain_openai import ChatOpenAI
    from langgraph.prebuilt import create_react_agent
    from tensorfeed.langchain import tensorfeed_tools

    agent = create_react_agent(ChatOpenAI(model="gpt-5.5"), tensorfeed_tools())
    agent.invoke({"messages": [("user", "What is happening with Anthropic today?")]})

This module imports ``langchain_core`` lazily so the base ``tensorfeed``
install stays stdlib-only.
"""

from __future__ import annotations

import json
from typing import Any

from .client import TensorFeed
from ._premium_tools import DESCRIPTIONS, FORMATTERS, safe_call


def _require_langchain() -> tuple[Any, Any]:
    try:
        from langchain_core.tools import StructuredTool
        from langchain_core.documents import Document
    except ImportError as e:  # pragma: no cover
        raise ImportError(
            "langchain_core is required for tensorfeed.langchain. "
            "Install with: pip install 'tensorfeed[langchain]'"
        ) from e
    return StructuredTool, Document


def _client(token: str | None) -> TensorFeed:
    return TensorFeed(token=token, user_agent="TensorFeed-LangChain/1.0")


# ── Tools ─────────────────────────────────────────────────────────


def TensorFeedNewsTool(token: str | None = None):
    """LangChain tool returning the latest AI news from /api/news."""
    StructuredTool, _ = _require_langchain()

    def _news(category: str | None = None, limit: int = 20) -> str:
        result = _client(token).news(category=category, limit=limit)
        items = result.get("articles", [])
        lines = []
        for a in items[:limit]:
            lines.append(
                f"- {a.get('title', '')} ({a.get('source', '')}, {a.get('publishedAt', '')}): {a.get('snippet', '')}"
            )
        return "\n".join(lines) if lines else "No articles."

    return StructuredTool.from_function(
        func=_news,
        name="tensorfeed_news",
        description=(
            "Get the latest AI news headlines and snippets from TensorFeed. "
            "Aggregated from 15+ sources, updated every 10 minutes. "
            "Optional `category` filter (research, tools, anthropic, openai, etc). "
            "Default limit 20, max 100."
        ),
    )


def TensorFeedStatusTool(token: str | None = None):
    """LangChain tool returning live AI service status from /api/status."""
    StructuredTool, _ = _require_langchain()

    def _status() -> str:
        result = _client(token).status()
        services = result.get("services", [])
        lines = []
        for s in services:
            name = s.get("name", "")
            status = s.get("status", "unknown")
            lines.append(f"- {name}: {status}")
        return "\n".join(lines) if lines else "No status data."

    return StructuredTool.from_function(
        func=_status,
        name="tensorfeed_status",
        description=(
            "Get the real-time operational status of major AI services "
            "(Claude, ChatGPT, Gemini, Mistral, Cohere, Replicate, Hugging Face). "
            "Polled every 5 minutes. Status: operational, degraded, down, or unknown."
        ),
    )


def TensorFeedRoutingTool(token: str | None = None):
    """LangChain tool returning the free top-1 model routing recommendation."""
    StructuredTool, _ = _require_langchain()

    def _routing(task: str = "general") -> str:
        result = _client(token).routing_preview(task=task)
        rec = result.get("recommendation") or {}
        if not rec:
            return "No routing recommendation available."
        return json.dumps(
            {
                "task": task,
                "model": rec.get("model"),
                "provider": rec.get("provider"),
                "blended_price": rec.get("blendedPrice"),
                "score": rec.get("score"),
            },
            indent=2,
        )

    return StructuredTool.from_function(
        func=_routing,
        name="tensorfeed_routing_preview",
        description=(
            "Get the top-1 recommended AI model for a given task. "
            "Task can be 'code', 'reasoning', 'creative', or 'general'. "
            "Free preview tier; for the full ranked list with score breakdown "
            "use the paid /api/premium/routing endpoint."
        ),
    )


def TensorFeedAttentionTool(token: str | None = None):
    """LangChain tool returning the live AI Attention Index."""
    StructuredTool, _ = _require_langchain()

    def _attention() -> str:
        result = _client(token).attention()
        rows = result.get("providers", [])[:10]
        lines = ["AI Attention Index (live):"]
        for p in rows:
            lines.append(
                f"  #{p['rank']:>2} {p['name']:<14} {p['attention_score']:>5.1f}  "
                f"({p['news_24h']} news/24h, {p['trending_repos']} trending repos)"
            )
        return "\n".join(lines)

    return StructuredTool.from_function(
        func=_attention,
        name="tensorfeed_attention",
        description=(
            "Get the live AI Attention Index ranking AI providers by news volume, "
            "GitHub trending repos, and inbound agent traffic. Score is 0-100, "
            "normalized within each response. Updates every 5 minutes."
        ),
    )


def TensorFeedHarnessesTool(token: str | None = None):
    """LangChain tool returning the cross-harness coding agent leaderboard."""
    StructuredTool, _ = _require_langchain()

    def _harnesses(benchmark: str = "swe_bench_verified") -> str:
        result = _client(token).harnesses()
        rows = [
            r
            for r in result.get("results", [])
            if isinstance(r.get("scores", {}).get(benchmark), (int, float))
        ]
        rows.sort(key=lambda r: r["scores"][benchmark], reverse=True)
        lines = [f"Coding harness leaderboard ({benchmark}):"]
        for r in rows[:15]:
            lines.append(
                f"  {r['scores'][benchmark]:>5.1f}  {r['harness']:<18} on {r['model']}"
            )
        return "\n".join(lines)

    return StructuredTool.from_function(
        func=_harnesses,
        name="tensorfeed_harnesses",
        description=(
            "Get the cross-harness coding agent leaderboard. "
            "Compares Claude Code, Cursor Agent, Codex CLI, Aider, OpenHands, "
            "Devin, Cline, Windsurf, Amp, Continue, Roo Code on agentic benchmarks. "
            "Pass `benchmark` as one of: swe_bench_verified (default), terminal_bench, "
            "aider_polyglot, swe_lancer."
        ),
    )


def tensorfeed_tools(
    token: str | None = None,
    *,
    include_premium: bool = False,
) -> list[Any]:
    """Convenience: the TensorFeed LangChain tools.

    Returns the five free tools by default. Pass
    ``include_premium=True`` to also attach the paid catalog (1 credit
    per call). The premium tools are safe to attach even without a
    token: called with no credits they return guidance instead of
    paying or raising, so an agent can still reason about them.

    Args:
        token: Optional bearer token for premium calls. Free tools
            ignore it.
        include_premium: Append the premium tools as well.
    """
    tools = [
        TensorFeedNewsTool(token),
        TensorFeedStatusTool(token),
        TensorFeedRoutingTool(token),
        TensorFeedAttentionTool(token),
        TensorFeedHarnessesTool(token),
    ]
    if include_premium:
        tools.extend(tensorfeed_premium_tools(token))
    return tools


# ── Premium tools (paid, 1 credit each, never auto-pay) ───────────────
#
# Each tool has an explicit typed signature so LangChain builds a clean
# argument schema for the model. The call is routed through the shared
# ``safe_call`` so the payment posture and result formatting are
# identical to the CrewAI binding and tested once.


def _premium_structured_tool(name: str, fn):
    StructuredTool, _ = _require_langchain()
    return StructuredTool.from_function(
        func=fn, name=name, description=DESCRIPTIONS[name]
    )


def TensorFeedWhatsNewTool(token: str | None = None):
    """Paid agent boot brief. 1 credit. Will not auto-pay."""

    def _whats_new(days: int = 1, news_limit: int = 10) -> str:
        return safe_call(
            lambda: _client(token).whats_new(days=days, news_limit=news_limit),
            FORMATTERS["tensorfeed_whats_new"],
        )

    return _premium_structured_tool("tensorfeed_whats_new", _whats_new)


def TensorFeedRoutingFullTool(token: str | None = None):
    """Paid full ranked routing. 1 credit. Will not auto-pay."""

    def _routing(
        task: str = "general",
        budget: float | None = None,
        top_n: int = 5,
    ) -> str:
        return safe_call(
            lambda: _client(token).routing(task=task, budget=budget, top_n=top_n),
            FORMATTERS["tensorfeed_routing"],
        )

    return _premium_structured_tool("tensorfeed_routing", _routing)


def TensorFeedCompareModelsTool(token: str | None = None):
    """Paid 2-5 model comparison. 1 credit. Will not auto-pay."""

    def _compare(ids: str) -> str:
        return safe_call(
            lambda: _client(token).compare_models(ids=ids),
            FORMATTERS["tensorfeed_compare_models"],
        )

    return _premium_structured_tool("tensorfeed_compare_models", _compare)


def TensorFeedCostProjectionTool(token: str | None = None):
    """Paid workload cost projection. 1 credit. Will not auto-pay."""

    def _cost(
        models: str,
        input_tokens_per_day: float,
        output_tokens_per_day: float,
        horizon: str | None = None,
    ) -> str:
        return safe_call(
            lambda: _client(token).cost_projection(
                models=models,
                input_tokens_per_day=input_tokens_per_day,
                output_tokens_per_day=output_tokens_per_day,
                horizon=horizon,
            ),
            FORMATTERS["tensorfeed_cost_projection"],
        )

    return _premium_structured_tool("tensorfeed_cost_projection", _cost)


def TensorFeedNewsSearchTool(token: str | None = None):
    """Paid full-text news search. 1 credit. Will not auto-pay."""

    def _search(
        q: str | None = None,
        from_date: str | None = None,
        to_date: str | None = None,
        provider: str | None = None,
        category: str | None = None,
        limit: int = 25,
    ) -> str:
        return safe_call(
            lambda: _client(token).news_search(
                q=q,
                from_date=from_date,
                to_date=to_date,
                provider=provider,
                category=category,
                limit=limit,
            ),
            FORMATTERS["tensorfeed_news_search"],
        )

    return _premium_structured_tool("tensorfeed_news_search", _search)


def TensorFeedProviderDeepDiveTool(token: str | None = None):
    """Paid one-call provider profile. 1 credit. Will not auto-pay."""

    def _deepdive(provider: str) -> str:
        return safe_call(
            lambda: _client(token).provider_deepdive(provider),
            FORMATTERS["tensorfeed_provider_deepdive"],
        )

    return _premium_structured_tool("tensorfeed_provider_deepdive", _deepdive)


def TensorFeedStatusLeaderboardTool(token: str | None = None):
    """Paid 90-day uptime leaderboard. 1 credit. Will not auto-pay."""

    def _leaderboard(
        from_date: str | None = None,
        to_date: str | None = None,
    ) -> str:
        return safe_call(
            lambda: _client(token).status_leaderboard(
                from_date=from_date, to_date=to_date
            ),
            FORMATTERS["tensorfeed_status_leaderboard"],
        )

    return _premium_structured_tool("tensorfeed_status_leaderboard", _leaderboard)


def tensorfeed_premium_tools(token: str | None = None) -> list[Any]:
    """The paid TensorFeed tool catalog (1 credit per call).

    Safe to attach without a token: each tool returns actionable
    guidance instead of paying or raising when no credits are
    available. None of these tools move funds.
    """
    return [
        TensorFeedWhatsNewTool(token),
        TensorFeedRoutingFullTool(token),
        TensorFeedCompareModelsTool(token),
        TensorFeedCostProjectionTool(token),
        TensorFeedNewsSearchTool(token),
        TensorFeedProviderDeepDiveTool(token),
        TensorFeedStatusLeaderboardTool(token),
    ]


# ── Document loader ───────────────────────────────────────────────


class TensorFeedNewsLoader:
    """LangChain document loader for TensorFeed news articles.

    Each article becomes a ``Document`` with ``page_content`` set to
    title + snippet, and metadata covering source, URL, categories,
    and publishedAt. Useful for RAG over recent AI news.

    Usage::

        from tensorfeed.langchain import TensorFeedNewsLoader

        docs = TensorFeedNewsLoader(category="research", limit=50).load()
    """

    def __init__(
        self,
        category: str | None = None,
        limit: int = 50,
        token: str | None = None,
    ) -> None:
        self.category = category
        self.limit = limit
        self.token = token

    def load(self) -> list[Any]:
        _, Document = _require_langchain()
        result = _client(self.token).news(category=self.category, limit=self.limit)
        docs = []
        for a in result.get("articles", []):
            docs.append(
                Document(
                    page_content=f"{a.get('title', '')}\n\n{a.get('snippet', '')}",
                    metadata={
                        "source": a.get("source", ""),
                        "source_domain": a.get("sourceDomain", ""),
                        "url": a.get("url", ""),
                        "published_at": a.get("publishedAt", ""),
                        "categories": a.get("categories", []),
                        "id": a.get("id", ""),
                    },
                )
            )
        return docs

    def lazy_load(self):
        for doc in self.load():
            yield doc

File: sdk/javascript/src/langchain.ts
/**
 * TensorFeed tools for LangChain.js (`@langchain/core`).
 *
 * Subpath entry point: `import { tensorfeedPremiumTools } from
 * 'tensorfeed/langchain'`. `@langchain/core` and `zod` are optional
 * peer dependencies (zod is already LangChain's own peer dep), so the
 * base `tensorfeed` install stays dependency-free; both are imported
 * lazily (via non-literal specifiers so the build does not require
 * them) and only when this module is used.
 *
 * Payment posture: these tools never move funds. Called without a
 * credit token they return actionable guidance instead of paying or
 * throwing. The invariant lives in ./premium-tools and is tested there.
 */

import { TensorFeed } from './index.js';
import {
  DESCRIPTIONS,
  FORMATTERS,
  INVOKERS,
  PREMIUM_TOOL_NAMES,
  safeCall,
} from './premium-tools.js';

/** Minimal structural view of the zod surface we use (no any). */
interface ZodType {
  optional(): ZodType;
  min(n: number): ZodType;
  max(n: number): ZodType;
  int(): ZodType;
  describe(d: string): ZodType;
}
interface Zod {
  object(shape: Record<string, ZodType>): ZodType;
  string(): ZodType;
  number(): ZodType;
  enum(values: string[]): ZodType;
}
interface LangchainToolsModule {
  tool: (
    fn: (args: Record<string, unknown>) => Promise<string>,
    config: { name: string; description: string; schema: unknown },
  ) => unknown;
}
interface ZodModule {
  z?: Zod;
  default?: Zod;
}

async function loadDeps(): Promise<{ tool: LangchainToolsModule['tool']; z: Zod }> {
  const toolsSpec: string = '@langchain/core/tools';
  const zodSpec: string = 'zod';
  let toolMod: LangchainToolsModule;
  let zodMod: ZodModule;
  try {
    toolMod = await import(toolsSpec);
  } catch {
    throw new Error(
      '@langchain/core is required for tensorfeed/langchain. Install it ' +
        'with: npm install @langchain/core zod',
    );
  }
  try {
    zodMod = await import(zodSpec);
  } catch {
    throw new Error('zod is required for tensorfeed/langchain. Install it with: npm install zod');
  }
  const z = zodMod.z ?? zodMod.default;
  if (!z) throw new Error('zod module did not expose a usable `z` export.');
  return { tool: toolMod.tool, z };
}

function schemas(z: Zod): Record<string, ZodType> {
  return {
    tensorfeed_whats_new: z.object({
      days: z.number().int().min(1).max(7).optional(),
      news_limit: z.number().int().min(1).max(25).optional(),
    }),
    tensorfeed_routing: z.object({
      task: z.enum(['code', 'reasoning', 'creative', 'general']).optional(),
      budget: z.number().optional(),
      top_n: z.number().int().min(1).max(10).optional(),
    }),
    tensorfeed_compare_models: z.object({
      ids: z.string().describe('2 to 5 model ids or names, comma-separated'),
    }),
    tensorfeed_cost_projection: z.object({
      models: z.string().describe('1 to 10 model ids or names, comma-separated'),
      input_tokens_per_day: z.number(),
      output_tokens_per_day: z.number(),
      horizon: z.enum(['daily', 'weekly', 'monthly', 'yearly']).optional(),
    }),
    tensorfeed_news_search: z.object({
      q: z.string().optional(),
      from_date: z.string().describe('YYYY-MM-DD UTC').optional(),
      to_date: z.string().describe('YYYY-MM-DD UTC').optional(),
      provider: z.string().optional(),
      category: z.string().optional(),
      limit: z.number().int().min(1).max(100).optional(),
    }),
    tensorfeed_provider_deepdive: z.object({
      provider: z.string().describe('Provider id or display name'),
    }),
    tensorfeed_status_leaderboard: z.object({
      from_date: z.string().describe('YYYY-MM-DD UTC').optional(),
      to_date: z.string().describe('YYYY-MM-DD UTC').optional(),
    }),
  };
}

export interface PremiumToolsOptions {
  /** Bearer token for premium calls. Omit to get guidance-only tools. */
  token?: string;
}

/**
 * The paid TensorFeed catalog as LangChain.js structured tools (1
 * credit per call). Returns an array ready to pass to an agent or
 * `bindTools`.
 *
 * Safe to attach without a token: each tool returns actionable guidance
 * instead of paying or throwing when no credits are available. None of
 * these tools move funds.
 *
 * Async because the optional deps are imported lazily:
 *
 *   const tools = await tensorfeedPremiumTools({ token });
 *   const agent = createReactAgent({ llm, tools });
 */
export async function tensorfeedPremiumTools(
  options?: PremiumToolsOptions,
): Promise<unknown[]> {
  const { tool, z } = await loadDeps();
  const client = new TensorFeed({
    token: options?.token,
    userAgent: 'TensorFeed-LangChainJS/1.0',
  });
  const schemaByName = schemas(z);

  return PREMIUM_TOOL_NAMES.map((name) =>
    tool(
      (args: Record<string, unknown>) =>
        safeCall(() => INVOKERS[name](client, args ?? {}), FORMATTERS[name]),
      {
        name,
        description: DESCRIPTIONS[name],
        schema: schemaByName[name],
      },
    ),
  );
}
